Including Data Integrity Test program in fabtests.

prankurgupta opened the pull request:
Currently, this program uses EP_MSG hardcoded in it.

Signed-off-by: Prankur Gupta <prankgup@cisco.com>

@goodell

pmmccorm commented:
If you are going to add options to the common set, then all tests need to support them. There are a couple examples in the tree of tests that have custom options: I think rdm_atomic is one.

I think this might make more sense as a patch to an existing test adding verification.

prankurgupta commented:
Removed the dependency from common structures, added custom option arguments.
We're looking at whether any other test would be suitable for this.

@pmmccorm @goodell

pmmccorm commented:
What about just including this with pingpong/msg_pingping.c ?

It looks like it is right now based off of an old version, might as well just add this capability to an existing test instead of: copy test, add minor changes, then rename-- this is how we end up with ten different ping-ping tests that all differ by ten or twenty lines..

shantonu commented:
I agree with @pmmccorm. We can add data integrity test as a part of existing tests as an option. Also, it'll be a good idea to add new tests to /scripts/runfabtests.sh.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+		case '?':
+		case 'h':
+			ft_csusage(argv[0], "Ping pong client and server using message endpoints.");
+			FT_PRINT_OPTS_USAGE("-c <count>", "allowed error count");
```
I'm not sure I see the value of the `allowed_error_count`.  EP_MSG and EP_RDM endpoints are supposed to be fully reliable, so it's a real problem as soon as there is a single instance of data corruption.

goodell commented:
@prankurgupta I haven't finished my line-by-line review yet, but will soon.

@prankurgupta: @shantonu's suggestion to add this to `scripts/runfabtests.sh` is a good one

@pmmccorm @shantonu: do you have a test you would recommend we add this to instead?  We deliberately avoided modifying `msg_pingpong.c` to add this support because we wanted the pingpong tests to serve as simpler examples of how to use the API (and they're already complicated enough as-is).

Over the longer-term the plan is to morph this program into checking for data integrity issues in EP_MSG, EP_RDM, and EP_DGRAM endpoints.  So while the name is accurate right now, it should be updated as the test is expanded.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+/*
+ * Copyright (c) 2013-2014 Intel Corporation.  All rights reserved.
```
Please add the Cisco 2015 copyright line here.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+	if (opts.machr)
+		show_perf_mr(opts.transfer_size, opts.iterations, &start, &end, 2, opts.argc, opts.argv);
+	else
+		show_perf(test_name, opts.transfer_size, opts.iterations, &start, &end, 2);
```
Do we need to `show_perf` or `show_perf_mr` at all?  The performance of this test is basically irrelevant.  It would be nice to emit some sort of output for each message size tested, but the performance numbers don't seem like the right thing to show.  Maybe just one line for each message size, indicating `   128... OK` or something like that.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+	}
+	if (detailed_error) {
+                print_error = 1;
+        }
```
spacing looks off here

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+	printf("%s = ", buffer);
+	for(j=0; j<size; j++) {
+		printf("%c", array[j]);
+        }
```
indenting looks inconsistent here... maybe you're mixing tabs and spaces?  The libfabric/fabtests style is essentially the Linux kernel style, which consistently uses hard tabs with an 8 character spacing.

Also, the spacing on the `for()` line doesn't match the rest of the project.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+
+	if (iter >= 0) {
+                msg_index = (iter*common_seed) % msg_char_length;
+        } else {
```
more inconsistent indenting, same goes for the rest of this function

shefty commented:
It should be possible to add data integrity checks into the existing test apps such that it doesn't result in making the code any more complex to read.  The functions that format and check the data are separated out.

We absolutely need to end the copy-paste-modify of these tests.  Updates are too time consuming.  Effort needs to go into making more of the code common.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+			}
+			return ret;
+		}
+	} while (ret == -FI_EAGAIN);
```
I know this was probably cribbed from `msg_pingpong.c`, but you might consider using `fi_cq_sread` in order to ditch the loop.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+		goto no_error;
+	}
+	if (iter < 0) {
+		printf("Test Name = %s, Sync test. Size = %d First_Error_byte = %d\n", 
```
The message should say either `first_error_byte` or `first error byte`.  The mixed case and `_` usage is weird.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+			test_name, size, first_error_byte);
+	} else {
+		printf("Test Name = %s, Iteration = %d Size = %d"
+			" First_Error_byte = %d\n", test_name, iter,
```
ditto here

pmmccorm commented:
What Sean said. This should be a twenty-line patch, not another near-identical copy of a pingpong test that adds one (simple) feature.

goodell commented on simple/msg_data_integrity.c:
```diff
@@ -0,0 +1,717 @@
+	if (iter < 0) {
+		next_size = transfer_size;
+	}
+	ret = fi_recv(ep, recv_buf, next_size, fi_mr_desc(recv_mr), 0, recv_buf);
```
do you still have an issue where the last recv posted here matches up with the recv posted in `ft_finalize`?

goodell commented:
@prankurgupta ok, I finished reviewing.  If you have any questions, feel free to reach out to me directly for clarification.

goodell commented:
@pmmccorm @shefty alright, we'll try to add this to `pingpong/msg_pingpong.c` in a way that is minimally intrusive.

shefty commented:
I'm going to close this pull request, and assume that a future PR will update the existing tests.  Please note that I do have plans to make the existing tests more consistent with each other and move common functionality into a shared location.  Hopefully that will allow the data verification to be picked up by most of the tests.
